src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/Collect.cs

```diff
@@ -0,0 +1,492 @@
+                bool checkAggregateNames = argType.CheckAggregateNames(datasourceType, args[i], errors, features, SupportsParamCoercion);
+                fValid = fValid && checkAggregateNames;
+
+                if (!itemType.IsValid)
+                {
+                    itemType = argType;
+                }
+                else
+                {
+                    var fUnionError = false;
+                    itemType = DType.Union(ref fUnionError, itemType, argType, useLegacyDateTimeAccepts: true, features);
+                    if (fUnionError)
+                    {
+                        errors.EnsureError(DocumentErrorSeverity.Severe, args[i], TexlStrings.ErrIncompatibleTypes);
+                        fValid = false;
+                    }
+                }
+
+                // We only support accessing entities in collections if the collection has only 1 argument that contributes to it's type
+                if (argc != 2 && itemType.ContainsDataEntityType(DPath.Root))
+                {
+                    fValid &= DropAllOfKindNested(ref itemType, errors, args[i], DKind.DataEntity);
+                }
+            }
+
+            Contracts.Assert(!itemType.IsValid || itemType.IsTable);
+            collectedType = itemType.IsValid ? itemType : DType.EmptyTable;
+            return fValid;
+        }
+
+        // Typecheck an invocation of Collect.
+        public override bool CheckTypes(CheckTypesContext context, TexlNode[] args, DType[] argTypes, IErrorContainer errors, out DType returnType, out Dictionary<TexlNode, DType> nodeToCoercedTypeMap)
+        {
+            Contracts.AssertValue(args);
+            Contracts.AssertAllValues(args);
+            Contracts.AssertValue(argTypes);
+            Contracts.Assert(args.Length == argTypes.Length);
+            Contracts.AssertValue(errors);
+            Contracts.Assert(MinArity <= args.Length && args.Length <= MaxArity);
+
+            var fValid = base.CheckTypes(context, args, argTypes, errors, out returnType, out nodeToCoercedTypeMap);
+
+            // Need a collection for the 1st arg
+            DType collectionType = argTypes[0];
+            if (!collectionType.IsTable)
+            {
+                errors.EnsureError(DocumentErrorSeverity.Severe, args[0], TexlStrings.ErrInvalidArgs_Func, Name);
+                fValid = false;
+            }
+
+            DType collectedType = null;
+
+            // Get the unified collected type on the RHS. This will generate appropriate
+            // document errors for invalid arguments such as unsupported aggregate types.
+            if (context.Features.PowerFxV1CompatibilityRules && !context.AnalysisMode)
+            {
+                fValid &= TryGetUnifiedCollectedTypeV1(args, argTypes, errors, context.Features, out collectedType);
+            }
+            else
+            {
+                fValid &= TryGetUnifiedCollectedTypeCanvas(args, argTypes, errors, context.Features, out collectedType);

```

Comment:
>TryGetUnifiedCollectedTypeCanvas [](http://example.com/codeflow?start=26&length=32)

See other comment for naming suggestion. Other hosts may, in the future, allow for type expansion as well